src/cursor/cur_version.c

```diff
@@ -0,0 +1,611 @@
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_ITEM(data);
+    WT_DECL_ITEM(metadata);
+    WT_DECL_PACK_VALUE(pv);
+    WT_DECL_RET;
+    WT_PACK pack;
+    WT_SESSION_IMPL *session;
+    const uint8_t *p, *end;
+    va_list ap;
+
+    CURSOR_API_CALL(cursor, session, get_value, NULL);
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    if (F_ISSET(cursor, WT_CURSTD_RAW)) {
+        /* Extract metadata and value separately as raw data. */
+        metadata = va_arg(ap, WT_ITEM *);
+        metadata->data = cursor->value.data;
+        metadata->size = cursor->value.size;
+        data = va_arg(ap, WT_ITEM *);
+        data->data = table_cursor->value.data;
+        data->size = table_cursor->value.size;
+    } else {
+        /*
+         * Unpack the metadata. We cannot use the standard get value function here because variable
+         * arguments cannot be partially extracted by different function calls.
+         */
+        p = (uint8_t *)cursor->value.data;
+        end = p + cursor->value.size;
+
+        WT_ERR(__pack_init(session, &pack, WT_VERSION_CURSOR_METADATA_FORMAT));
+        while ((ret = __pack_next(&pack, &pv)) == 0) {
+            WT_ERR(__unpack_read(session, &pv, &p, (size_t)(end - p)));
+            WT_UNPACK_PUT(session, pv, ap);
+        }
+        WT_ERR_NOTFOUND_OK(ret, false);
+
+        WT_ASSERT(session, p <= end);
+        WT_ERR(__wt_cursor_get_valuev(table_cursor, ap));
+    }
+
+err:
+    va_end(ap);
+    API_END_RET(session, ret);
+}
+
+/*
+ * __curversion_next_int --
+ *     internal implementation for version cursor next
+ */
+static int
+__curversion_next_int(WT_SESSION_IMPL *session, WT_CURSOR *cursor)
+{
+    WT_CURSOR *hs_cursor, *table_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_ITEM(hs_value);
+    WT_DECL_ITEM(key);
+    WT_DECL_RET;
+    WT_PAGE *page;
+    WT_TIME_WINDOW *twp;
+    WT_UPDATE *upd, *tombstone;
+    wt_timestamp_t durable_start_ts, durable_stop_ts, stop_ts;
+    uint64_t stop_txn, hs_upd_type;
+    uint32_t raw;
+    uint8_t *p, version_prepare_state;
+    bool upd_found;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    hs_cursor = version_cursor->hs_cursor;
+    table_cursor = version_cursor->table_cursor;
+    cbt = (WT_CURSOR_BTREE *)table_cursor;
+    page = cbt->ref->page;
+    twp = NULL;
+    upd_found = false;
+    /* Temporarily clear the raw flag. We need to pack the data according to the format. */
+    raw = F_MASK(cursor, WT_CURSTD_RAW);
+    F_CLR(cursor, WT_CURSTD_RAW);
+
+    /* The cursor should be positioned, otherwise the next call will fail. */
+    if (!F_ISSET(table_cursor, WT_CURSTD_KEY_INT)) {
+        WT_IGNORE_RET(__wt_msg(
+          session, "WT_ROLLBACK: rolling back version_cursor->next due to no initial position"));
+        WT_ERR(WT_ROLLBACK);
+    }
+
+    upd = version_cursor->next_upd;
```

Comment:
Why can't this update have been discarded as obsolete between calls, that is, why won't `upd->txnid` drop core or hit garbage memory?